packages/core/src/lib/base.ts

```diff
@@ -0,0 +1,319 @@
+    }
+
+    await this.getBackend().storeContext(context);
+  }
+
+  /** Returns the current backend. */
+  protected getBackend(): B {
+    return this.backend;
+  }
+
+  /** Resolves all currently known breadcrumbs. */
+  protected async getBreadcrumbs(): Promise<Breadcrumb[]> {
+    if (!this.breadcrumbs) {
+      this.breadcrumbs = await this.getBackend().loadBreadcrumbs();
+    }
+
+    return this.breadcrumbs;
+  }
+
+  /**
+   * Sends an event (either error or message) to Sentry.
+   *
+   * This also adds breadcrumbs and context information to the event. However,
+   * platform specific meta data (such as the User's IP address) must be added
+   * by the SDK implementor.
+   *
+   * The returned event status offers clues to whether the event was sent to
+   * Sentry and accepted there. If the {@link Options.shouldSend} hook returns
+   * `false`, the status will be {@link SendStatus.Skipped}.If the rate limit
+   * was exceeded, the status will be {@link SendStatus.RateLimit}.
+   *
+   * @param event The event to send to Sentry.
+   * @returns A Promise that resolves with the event status.
+   */
+  protected async sendEvent(event: SentryEvent): Promise<SendStatus> {
+    if (!this.isEnabled()) {
+      return SendStatus.Skipped;
+    }
+
+    const prepared = await this.prepareEvent(event);
+    const { shouldSend, beforeSend, afterSend } = this.getOptions();
+    if (shouldSend && !shouldSend(prepared)) {
+      return SendStatus.Skipped;
+    }
+
+    const finalEvent = beforeSend ? beforeSend(prepared) : prepared;
+    const code = await this.getBackend().sendEvent(finalEvent);
+    const status = SendStatus.fromHttpCode(code);
+
+    // TODO: Handle rate limits and maintain a queue. For now, we require SDK
+    // implementors to override this method and handle it themselves.
+
+    if (afterSend) {
+      afterSend(finalEvent, status);
+    }
+
+    return status;
+  }
+
+  /** Determines whether this SDK is enabled and a valid DSN is present. */
+  protected isEnabled(): boolean {
+    return this.getOptions().enabled !== false && this.dsn !== undefined;
+  }
+
+  /**
+   * Adds common information to events, such as breadcrumbs, context, release
+   * and environment information.
+   *
+   * @param event The original event.
+   * @returns A new event with more information.
+   */
+  private async prepareEvent(event: SentryEvent): Promise<SentryEvent> {
+    const {
+      environment,
+      maxBreadcrumbs = MAX_BREADCRUMBS,
+      release,
+    } = this.getOptions();
+
+    const prepared = { ...event };
+    if (environment !== undefined) {
+      prepared.environment = environment;
+    }
+    if (release !== undefined) {
+      prepared.release = release;
+    }
+
+    const breadcrumbs = await this.getBreadcrumbs();
+    if (breadcrumbs.length > 0 && maxBreadcrumbs > 0) {
+      prepared.breadcrumbs = breadcrumbs.slice(-maxBreadcrumbs);
+    }
+
+    const context = await this.getContext();
```

Comment:
This is duplicated, but it's totally fine in this case I guess

Reply:
What is duplicated here?